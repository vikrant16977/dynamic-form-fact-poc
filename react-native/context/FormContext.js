import React, { createContext, useState, useEffect, useRef } from "react";
import axios from "axios";

export const FormContext = createContext();

export const FormProvider = ({ children }) => {
  const [forms, setForms] = useState([]);
  const [selectedFormId, setSelectedFormId] = useState(null);
  const [loading, setLoading] = useState(true);
  const intervalRef = useRef(null);

  const fetchForms = async () => {
    try {
      const response = await axios.get(
        "https://dynamic-form-mobile-noktwd7ra-vikrant-ahers-projects.vercel.app/api/forms",
        {
          headers: {
            "User-Agent": "Mozilla/5.0",
          },
        }
      );
      const data = response.data;
      console.log("Fetched data:", data);

      if (!data || !Array.isArray(data.value)) {
        console.error("Unexpected API response:", data);
        setForms([]);
        setLoading(false);
        return;
      }

      const formsFetched = [];

      for (const item of data.value) {
        try {
          let parsedSchema;

          if (typeof item.schema === "string") {
            let cleanStr = item.schema;

            if (cleanStr.startsWith('"') && cleanStr.endsWith('"')) {
              cleanStr = cleanStr.slice(1, -1);
            }

            cleanStr = cleanStr.replace(/\\"/g, '"');

            parsedSchema = JSON.parse(cleanStr);
          } else {
            parsedSchema = item.schema;
          }

          if (Array.isArray(parsedSchema)) {
            parsedSchema.forEach((form) => {
              formsFetched.push({ ...form, id: item.ID });
            });
          } else {
            formsFetched.push({ ...parsedSchema, id: item.ID });
          }
        } catch (e) {
          console.error("Failed to parse form schema for item:", item);
          console.error("Error:", e);
        }
      }

      setForms(formsFetched);
      if (formsFetched.length > 0) setSelectedFormId(formsFetched[0].id);
    } catch (error) {
      console.error("Failed to fetch forms:", error);
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    fetchForms();

    intervalRef.current = setInterval(() => {
      fetchForms();
    }, 240000);

    return () => {
      if (intervalRef.current) {
        clearInterval(intervalRef.current);
      }
    };
  }, []);

  const addNewForm = (title) => {
    const newForm = { id: Date.now(), title, sections: [] };
    setForms((f) => [...f, newForm]);
    setSelectedFormId(newForm.id);
  };

  const addSectionToForm = (formId, section) => {
    setForms((f) =>
      f.map((fm) =>
        fm.id === formId ? { ...fm, sections: [...fm.sections, section] } : fm
      )
    );
  };

  const addQuestionToForm = (formId, sectionId, question) => {
    setForms((f) =>
      f.map((fm) => {
        if (fm.id !== formId) return fm;
        return {
          ...fm,
          sections: fm.sections.map((sec) =>
            sec.id === sectionId
              ? { ...sec, questions: [...sec.questions, question] }
              : sec
          ),
        };
      })
    );
  };

  const updateSelectedFormId = (id) => setSelectedFormId(id);
  const getSelectedForm = () => forms.find((fm) => fm.id === selectedFormId);

  return (
    <FormContext.Provider
      value={{
        forms,
        addNewForm,
        addSectionToForm,
        addQuestionToForm,
        updateSelectedFormId,
        getSelectedForm,
        selectedFormId,
        setForms,
        loading,
      }}
    >
      {children}
    </FormContext.Provider>
  );
};
