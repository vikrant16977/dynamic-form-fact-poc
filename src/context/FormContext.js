// src/context/FormContext.js
import React, { createContext, useState, useEffect } from "react";
import { v4 as uuidv4 } from 'uuid';

export const FormContext = createContext();

export const FormProvider = ({ children }) => {
  const [forms, setForms] = useState([]);
  const [activeFormId, setActiveFormId] = useState(null);

  // Load from localStorage
  useEffect(() => {
    const stored = localStorage.getItem("forms");
    if (stored) {
      setForms(JSON.parse(stored));
    }
  }, []);

  // Save to localStorage
  useEffect(() => {
    localStorage.setItem("forms", JSON.stringify(forms));
  }, [forms]);

  const createNewForm = (title) => {
    const newForm = { id: uuidv4(), title, fields: [] };
    setForms([...forms, newForm]);
    setActiveFormId(newForm.id);
  };

  const addFieldToForm = (formId, field) => {
    setForms(forms.map(form =>
      form.id === formId
        ? { ...form, fields: [...form.fields, field] }
        : form
    ));
  };

  const getActiveForm = () => forms.find(form => form.id === activeFormId);

  return (
    <FormContext.Provider value={{
      forms,
      activeFormId,
      createNewForm,
      addFieldToForm,
      setActiveFormId,
      getActiveForm
    }}>
      {children}
    </FormContext.Provider>
  );
};
