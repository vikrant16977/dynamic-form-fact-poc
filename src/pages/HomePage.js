import React, { useContext, useState } from "react";
import { useNavigate } from "react-router-dom";
import { FormContext } from "../context/FormContext";
import {
  Container,
  Header,
  Segment,
  Button,
  Dropdown,
  Grid,
} from "semantic-ui-react";
import HeaderBar from "../components/HeaderBar";

const HomePage = () => {
  const { forms, updateSelectedFormId } = useContext(FormContext);
  const [selectedForm, setSelectedForm] = useState(null);
  const navigate = useNavigate();

  const handleGoToEdit = () => {
    if (!selectedForm) return;
    updateSelectedFormId(selectedForm);
    navigate("/admin");
  };

  return (
    <>
      <HeaderBar />
      <Container style={{ marginTop: "15rem" }}>
        <Header as="h2" textAlign="center">
          FACT Application - Main Menu
        </Header>
        <Grid stackable columns={2} divided>
          <Grid.Column>
            <Segment>
              <Header as="h3">Create a New Audit Type</Header>
              <Button
              style={{ marginTop: "4.8rem" }}
                primary
                fluid
                onClick={() => navigate("/admin")}
              >
                Go to Audit Type Builder
              </Button>
            </Segment>
          </Grid.Column>

          <Grid.Column>
            <Segment>
              <Header as="h3">Edit Existing Audit Type</Header>
              {forms.length > 0 ? (
                <>
                  <Dropdown
                    placeholder="Select an Audit Type"
                    fluid
                    selection
                    options={forms.map((f) => ({
                      key: f.id,
                      text: f.title,
                      value: f.id,
                    }))}
                    value={selectedForm}
                    onChange={(e, { value }) => setSelectedForm(value)}
                    style={{ marginBottom: "1rem" }}
                  />
                  <Button
                    secondary
                    fluid
                    disabled={!selectedForm}
                    onClick={handleGoToEdit}
                  >
                    Go & Edit Audit Type
                  </Button>
                </>
              ) : (
                <Header as="h4" color="grey">
                  No Audit Type available. Create one first.
                </Header>
              )}
            </Segment>
          </Grid.Column>
        </Grid>
      </Container>
    </>
  );
};

export default HomePage;