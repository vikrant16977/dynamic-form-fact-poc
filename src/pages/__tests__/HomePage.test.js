import React from "react";
import { render, screen, fireEvent, within } from "@testing-library/react";
import HomePage from "../HomePage";
import { MemoryRouter } from "react-router-dom";
import { FormContext } from "../../context/FormContext";

// Mock HeaderBar to avoid rendering issues or image imports
jest.mock("../../components/HeaderBar", () => () => <div data-testid="header-bar" />);

const mockNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}));

const defaultForms = [
  { id: 1, title: "Form 1" },
  { id: 2, title: "Form 2" },
];

function renderWithContext(
  { forms = defaultForms, updateSelectedFormId = jest.fn(), addNewForm = jest.fn() } = {}
) {
  return render(
    <FormContext.Provider value={{ forms, updateSelectedFormId, addNewForm }}>
      <MemoryRouter>
        <HomePage />
      </MemoryRouter>
    </FormContext.Provider>
  );
}

describe("HomePage", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    window.localStorage.clear();
  });

  it("renders HeaderBar and main UI elements", () => {
    renderWithContext();
    expect(screen.getByTestId("header-bar")).toBeInTheDocument();
    expect(screen.getByRole("heading", { name: /create a new form/i })).toBeInTheDocument();
    expect(screen.getByPlaceholderText(/enter form name/i)).toBeInTheDocument();
    expect(screen.getByText(/select pre-existing forms/i)).toBeInTheDocument();
    expect(screen.getByText(/discard/i)).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /create/i })).toBeInTheDocument();
    expect(screen.getAllByText(/or/i).length).toBeGreaterThan(0);
  });

  it("enables and disables the Create button correctly", () => {
    renderWithContext();
    const createBtn = screen.getByRole("button", { name: /create/i });
    expect(createBtn).toBeDisabled();

    const input = screen.getByPlaceholderText(/enter form name/i);
    fireEvent.change(input, { target: { value: "Test Form" } });
    expect(createBtn).not.toBeDisabled();

    fireEvent.change(input, { target: { value: "" } });
    expect(createBtn).toBeDisabled();
  });

  it("handles form name input and calls addNewForm & navigate on Create", () => {
    const addNewForm = jest.fn();
    renderWithContext({ addNewForm });
    const input = screen.getByPlaceholderText(/enter form name/i);
    fireEvent.change(input, { target: { value: "My New Form" } });
    const createBtn = screen.getByRole("button", { name: /create/i });
    fireEvent.click(createBtn);
    expect(addNewForm).toHaveBeenCalledWith("My New Form");
    expect(window.localStorage.getItem("newForm")).toBe("true");
    expect(mockNavigate).toHaveBeenCalledWith("/admin");
  });

  it("does not call addNewForm for empty or whitespace form name", () => {
    const addNewForm = jest.fn();
    renderWithContext({ addNewForm });
    const input = screen.getByPlaceholderText(/enter form name/i);
    fireEvent.change(input, { target: { value: "   " } });
    const createBtn = screen.getByRole("button", { name: /create/i });
    fireEvent.click(createBtn);
    expect(addNewForm).not.toHaveBeenCalled();
    expect(mockNavigate).not.toHaveBeenCalled();
  });

  it("shows dropdown options and handles selecting existing form", () => {
    renderWithContext();
    const dropdown = screen.getByText(/select pre-existing forms/i);
fireEvent.click(dropdown);

const menu = document.querySelector('.visible.menu.transition');
expect(menu).toBeTruthy(); // makes sure we found it
const option = within(menu).getByText("Form 1");
fireEvent.click(option);

    // Once selected, Edit button should appear
    expect(screen.getByRole("button", { name: /edit/i })).toBeInTheDocument();
  });

  it("calls updateSelectedFormId and navigates when Edit is clicked", () => {
    const updateSelectedFormId = jest.fn();
    renderWithContext({ updateSelectedFormId });
    const dropdown = screen.getByText(/select pre-existing forms/i);
fireEvent.click(dropdown);

const menu = document.querySelector('.visible.menu.transition');
expect(menu).toBeTruthy(); // makes sure we found it
const option = within(menu).getByText("Form 1");
fireEvent.click(option);

    const editBtn = screen.getByRole("button", { name: /edit/i });
    fireEvent.click(editBtn);
    expect(updateSelectedFormId).toHaveBeenCalledWith(1);
    expect(window.localStorage.getItem("newForm")).toBe("false");
    expect(mockNavigate).toHaveBeenCalledWith("/admin");
  });

  it("disables Edit button if no form is selected and input is empty", () => {
    renderWithContext();
    expect(screen.getByRole("button", { name: /create/i })).toBeDisabled();
  });

  it("resets input and dropdown selection on Discard", () => {
    renderWithContext();
    const input = screen.getByPlaceholderText(/enter form name/i);
    fireEvent.change(input, { target: { value: "Some Form" } });
    const discardBtn = screen.getByText(/discard/i);
    fireEvent.click(discardBtn);
    expect(input.value).toBe("");
    // Dropdown should reset to null value (undefined), so the Create button should be disabled
    expect(screen.getByRole("button", { name: /create/i })).toBeDisabled();
  });

  it("renders correctly with empty forms array", () => {
    renderWithContext({ forms: [] });
    expect(screen.getByText(/select pre-existing forms/i)).toBeInTheDocument();
    // No dropdown options
    fireEvent.click(screen.getByText(/select pre-existing forms/i));
    expect(screen.queryByText("Form 1")).not.toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { asFragment } = renderWithContext();
    expect(asFragment()).toMatchSnapshot();
  });
});