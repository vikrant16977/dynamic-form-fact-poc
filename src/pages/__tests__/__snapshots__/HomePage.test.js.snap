// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage matches snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="header-bar"
  />
  <div
    class="ui text container"
    style="margin-top: 10rem;"
  >
    <div
      class="ui raised very padded segment"
      style="box-shadow: 0 1px 6px rgba(0,0,0,0.1);"
    >
      <h3
        class="ui header"
      >
        Create a New Form
      </h3>
      <div
        class="ui fluid input"
        style="margin-bottom: 2rem; margin-top: 1rem;"
      >
        <input
          placeholder="Enter Form Name"
          type="text"
          value=""
        />
      </div>
      <div
        class="ui horizontal divider"
      >
        Or
      </div>
      <div
        aria-expanded="false"
        class="ui fluid selection dropdown"
        role="listbox"
        style="margin: 2rem 0px;"
        tabindex="0"
      >
        <div
          aria-atomic="true"
          aria-live="polite"
          class="divider default text"
          role="alert"
        >
          Select pre-existing Forms
        </div>
        <i
          aria-hidden="true"
          class="dropdown icon"
        />
        <div
          class="menu transition"
        >
          <div
            aria-checked="false"
            aria-selected="true"
            class="selected item"
            role="option"
            style="pointer-events: all;"
          >
            <span
              class="text"
            >
              Form 1
            </span>
          </div>
          <div
            aria-checked="false"
            aria-selected="false"
            class="item"
            role="option"
            style="pointer-events: all;"
          >
            <span
              class="text"
            >
              Form 2
            </span>
          </div>
        </div>
      </div>
      <div
        class="ui stackable two column grid"
      >
        <div
          class="column"
        >
          <button
            class="ui fluid button"
          >
            Discard
          </button>
        </div>
        <div
          class="column"
        >
          <button
            class="ui fluid disabled button"
            disabled=""
            style="background-color: rgb(17, 50, 158); color: white;"
            tabindex="-1"
          >
            Create
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
